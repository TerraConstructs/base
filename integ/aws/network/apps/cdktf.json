{
  "//": "Dummy file to satisfy the bundling findUp logic"
}
