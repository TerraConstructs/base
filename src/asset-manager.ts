import {
  DockerImageAssetLocation,
  DockerImageAssetSource,
  FileAssetLocation,
  FileAssetSource,
} from "./assets";

/**
 * Interface for managing assets (files and Docker images) in infrastructure stacks.
 *
 * This interface abstracts asset management to avoid circular dependencies between
 * stack implementations and storage constructs.
 */
export interface IAssetManager {
  /**
   * Register a file asset and return its location details.
   *
   * @param asset The file asset source information
   * @returns Location details for the uploaded file asset
   */
  addFileAsset(asset: FileAssetSource): FileAssetLocation;

  /**
   * Register a Docker image asset and return its location details.
   *
   * @param asset The Docker image asset source information
   * @returns Location details for the built and pushed Docker image
   */
  addDockerImageAsset(asset: DockerImageAssetSource): DockerImageAssetLocation;
}
