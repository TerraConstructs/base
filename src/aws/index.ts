// core
// export * from "./tag-aspect";
// export * from "./tag-manager";
export * from "./arn";
export * from "./aws-construct";
export * from "./aws-stack";
export * from "./aws-tags";
export * from "./log-retention";

// core JSII compatible generated interfaces
export * from "./provider-config.generated";

// classes
export * as iam from "./iam";
export * as encryption from "./encryption";
export * as compute from "./compute";
export * as network from "./network";
export * as storage from "./storage";
export * as notify from "./notify";
export * as edge from "./edge";
export * as cloudwatch from "./cloudwatch";
