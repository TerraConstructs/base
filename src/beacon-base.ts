import {
  Lazy,
  TerraformResource,
  TerraformElement,
  TerraformOutput,
  Aspects,
  IAspect,
} from "cdktf";
import { Construct, IConstruct } from "constructs";
import { SpecBase } from ".";

export interface BeaconProps {
  /**
   * The friendly name for beacon resources
   *
   * @default - `environmentName-id`
   */
  readonly friendlyName?: string;

  /**
   * Whether to register Terraform outputs for this beacon
   *
   * @default false
   */
  readonly registerOutputs?: boolean;

  /**
   * Optional override for the outputs name
   *
   * @default id
   */
  readonly outputName?: string;
}

export interface IBeacon {
  /**
   * Environment Name passed in from the CLI
   */
  readonly environmentName: string;
  readonly gridUUID: string;
  readonly outputs: Record<string, any>;
}

type TaggableConstruct = IConstruct & {
  tags?: { [key: string]: string };
  tagsInput?: { [key: string]: string };
};

function isTaggableConstruct(x: IConstruct): x is TaggableConstruct {
  return (
    TerraformResource.isTerraformResource(x) && "tags" in x && "tagsInput" in x
  );
}

const GRID_TAG_PREFIX = "et:grid";

// Add Grid Tags to all Beacon resources
export class GridTags implements IAspect {
  constructor(private tagsToAdd: Record<string, string>) {}
  visit(node: IConstruct) {
    if (isTaggableConstruct(node)) {
      // https://developer.hashicorp.com/terraform/cdktf/concepts/aspects
      const currentTags = node.tagsInput || {};
      node.tags = { ...this.tagsToAdd, ...currentTags };
    }
  }
}

/**
 * Base class for all Beacons
 *
 * Allows a Beacon to lazily register its outputs with its parent Spec
 */
export abstract class BeaconBase extends TerraformElement implements IBeacon {
  /**
   * The name under which the outputs are registered in the parent Scope
   */
  public readonly outputName: string;

  /**
   * Beacon friendly name
   */
  public readonly friendlyName: string;

  /**
   * Beacon unique grid identifier
   */
  public get gridUUID(): string {
    return SpecBase.ofBeacon(this).gridUUID;
  }

  /**
   * Environment Name passed in from the CLI
   */
  public get environmentName(): string {
    return SpecBase.ofBeacon(this).environmentName;
  }

  /**
   * Outputs to register with the parent Scope or undefined if there are no outputs
   */
  public abstract get outputs(): Record<string, any>;

  constructor(scope: Construct, id: string, props: BeaconProps) {
    super(scope, id);
    this.outputName = props.outputName || `${id}Outputs`;
    this.friendlyName = props.friendlyName || `${this.environmentName}-${id}`;

    Aspects.of(this) // Add Grid tags to every resource defined within.
      .add(
        new GridTags({
          [`${GRID_TAG_PREFIX}:EnvironmentName`]: this.environmentName,
          [`${GRID_TAG_PREFIX}:UUID`]: this.gridUUID,
          Name: this.friendlyName,
        }),
      );
    const registerOutputs = props.registerOutputs || false;
    if (registerOutputs) {
      new TerraformOutput(scope, this.outputName, {
        staticId: true,
        description: `Outputs for ${this.friendlyName}`,
        value: Lazy.anyValue({ produce: () => this.outputs || null }),
      });
    }
  }
}
