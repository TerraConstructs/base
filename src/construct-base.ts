import {
  Lazy,
  TerraformResource,
  TerraformElement,
  TerraformOutput,
  TerraformMetaArguments,
  Aspects,
  IAspect,
} from "cdktf";
import { Construct, IConstruct } from "constructs";
import { StackBase } from "./stack-base";

export interface TerraConstructProps extends TerraformMetaArguments {
  /**
   * The friendly name for TerraConstruct resources
   *
   * @default - `environmentName-id`
   */
  readonly friendlyName?: string;

  /**
   * Whether to register Terraform outputs for this TerraConstruct
   *
   * @default false
   */
  readonly registerOutputs?: boolean;

  /**
   * Optional override for the outputs name
   *
   * @default id
   */
  readonly outputName?: string;
}

export interface ITerraConstruct extends IConstruct {
  /**
   * Environment Name passed in from the CLI
   */
  readonly environmentName: string;
  readonly gridUUID: string;
  readonly outputs: Record<string, any>;
}

// TODO: this is aws specific, should be moved to aws module
export type TaggableConstruct = TerraformResource & {
  tags?: { [key: string]: string };
  tagsInput?: { [key: string]: string };
};

// TODO: this is aws specific, should be moved to aws module
export function isTaggableTerraformResource(
  x: IConstruct,
): x is TaggableConstruct {
  return (
    TerraformResource.isTerraformResource(x) && "tags" in x && "tagsInput" in x
  );
}

const GRID_TAG_PREFIX = "grid";

// Add Grid Tags to all TerraConstruct resources
export class GridTags implements IAspect {
  constructor(private tagsToAdd: Record<string, string>) {}
  visit(node: IConstruct) {
    if (isTaggableTerraformResource(node)) {
      // https://developer.hashicorp.com/terraform/cdktf/concepts/aspects
      const currentTags = node.tagsInput || {};
      // TODO: Bug - tagsToAdd are overwritten by currenTags
      node.tags = { ...this.tagsToAdd, ...currentTags };
    }
  }
}

/**
 * Base class for all TerraConstructs
 *
 * Allows a TerraConstruct to lazily register its outputs with its parent Stack
 */
export abstract class TerraConstructBase
  extends TerraformElement
  implements ITerraConstruct
{
  /**
   * Returns true if the construct was created by CDKTF, and false otherwise
   */
  public static isOwnedResource(construct: IConstruct): boolean {
    return construct.node.defaultChild
      ? TerraformResource.isTerraformResource(construct.node.defaultChild)
      : false;
  }

  /**
   * The name under which the outputs are registered in the parent Scope
   */
  public readonly outputName: string;

  /**
   * TerraConstruct friendly name
   */
  public readonly friendlyName: string;

  /**
   * TerraConstruct unique grid identifier
   */
  public get gridUUID(): string {
    return StackBase.ofTerraConstruct(this).gridUUID;
  }

  /**
   * Environment Name passed in from the CLI
   */
  public get environmentName(): string {
    return StackBase.ofTerraConstruct(this).environmentName;
  }

  /**
   * Outputs to register with the parent Scope or undefined if there are no outputs
   */
  public abstract get outputs(): Record<string, any>; // TODO: should be allowed to be undefined?

  constructor(
    scope: Construct,
    private readonly constructId: string,
    props: TerraConstructProps = {},
  ) {
    super(scope, constructId);
    this.outputName = props.outputName || `${constructId}Outputs`;
    this.friendlyName =
      props.friendlyName || `${this.environmentName}-${constructId}`;

    Aspects.of(this) // Add Grid tags to every resource defined within.
      .add(
        new GridTags({
          [`${GRID_TAG_PREFIX}:EnvironmentName`]: this.environmentName,
          [`${GRID_TAG_PREFIX}:UUID`]: this.gridUUID,
          Name: this.friendlyName,
        }),
      );
    const registerOutputs = props.registerOutputs || false;
    if (registerOutputs) {
      new TerraformOutput(scope, this.outputName, {
        staticId: true,
        description: `Outputs for ${this.friendlyName}`,
        value: Lazy.anyValue({ produce: () => this.outputs || null }),
      });
    }
  }

  // force usage of node.addDependency instead of passing TerraConstructs via dependsOn
  // Referring a TerraConstruct by fqn always triggers an error?
  public get fqn(): string {
    // try {
    //   return super.fqn;
    // } catch (e) {
    // ref: https://github.com/aws/constructs/blob/10.x/src/construct.ts#L345
    throw new Error(
      `Use Construct node.addDependency instead of passing TerraConstruct fqn ${this.constructor.name} ${this.constructId}`,
    );
    // }
  }
}
