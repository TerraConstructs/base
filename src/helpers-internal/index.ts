// https://github.com/aws/aws-cdk/blob/v2.175.1/packages/aws-cdk-lib/core/lib/helpers-internal/index.ts

export { md5hash } from "../private/md5";
