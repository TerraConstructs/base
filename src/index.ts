export * from "./construct-base";
export * from "./stack-base";
export * from "./duration";
export * from "./size";

// cloud specific exports
export * as aws from "./aws";
