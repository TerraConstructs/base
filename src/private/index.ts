export * from "./unique-resource-name";
export * from "./terraform-dependables-aspect";
