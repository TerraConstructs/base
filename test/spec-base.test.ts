import { App, Testing, TerraformResource, TerraformElement } from "cdktf";
import { Construct } from "constructs";
import "cdktf/lib/testing/adapters/jest";
import { SpecBase } from "../src";

const environmentName = "Test";
const gridUUID = "123e4567-e89b-12d3";
const gridBackendConfig = {
  address: "http://localhost:3000",
};
const terraformResourceType = "test_resource";

describe("SpecBase", () => {
  let app: App;
  let spec: MySpec;

  beforeEach(() => {
    app = Testing.app();
    spec = new MySpec(app, "TestSpec", {
      environmentName,
      gridUUID,
      gridBackendConfig,
    });
  });

  describe("TerraformDependencyAspect", () => {
    test("maps Construct dependencies to TerraformResource.dependsOn", () => {
      // GIVEN
      const simpleResource = new TerraformResource(spec, "SimpleResource", {
        terraformResourceType,
      });
      // a construct which is composed of nested resources
      const compositeResource = new CompositeResource(
        spec,
        "CompositeResource",
      );
      // a construct which adds nested resources during prepareStack
      const preSynthResource = new PreSynthResource(spec, "PreSynthResource");
      // a construct with 2 layers of nesting
      const deeplyNestedResource = new DeeplyNestedResource(
        spec,
        "DeeplyNestedResource",
      );

      // Dependables
      const directDependency = new TerraformResource(spec, "DirectDependency", {
        terraformResourceType,
      });
      const compositeDependency = new CompositeResource(
        spec,
        "CompositeDependency",
      );
      const presynthDependency = new PreSynthResource(
        spec,
        "PreSynthDependency",
      );

      // WHEN
      const expectedDependencies = new Array<string>();
      const resources = [
        simpleResource,
        compositeResource,
        preSynthResource,
        deeplyNestedResource,
      ];

      // Directly add dependencies to resources
      addDependencies(resources, directDependency);
      expectedDependencies.push(`${terraformResourceType}.DirectDependency`);

      // Add composite dependencies
      addDependencies(resources, compositeDependency);
      expectedDependencies.push(
        `${terraformResourceType}.CompositeDependency_NestedResource1_B2D8F1D4`,
        `${terraformResourceType}.CompositeDependency_NestedResource2_2E41AE93`,
      );

      // Add pre-synth dependencies
      addDependencies(resources, presynthDependency);
      expectedDependencies.push(
        `${terraformResourceType}.PreSynthDependency_NestedResource1_49A1A305`,
      );

      // THEN
      spec.prepareStack(); // required to add pre-synth resources
      const synthesized = Testing.synth(spec);
      // expect(synthesized).toMatchSnapshot();
      const template = JSON.parse(synthesized);
      expect(template).toMatchObject({
        resource: {
          [terraformResourceType]: {
            // direct resource depends on direct as well as nested resources
            // (including those added during prepareStack)
            SimpleResource: {
              depends_on: expectedDependencies,
            },
            // nested resources have the same dependencies through composite parent inheritance
            CompositeResource_NestedResource1_E176FFE6: {
              depends_on: expectedDependencies,
            },
            CompositeResource_NestedResource2_5A4D5ED7: {
              depends_on: expectedDependencies,
            },
            // pre-synth resources have the same dependencies through parent inheritance
            PreSynthResource_NestedResource1_A8EF732B: {
              depends_on: expectedDependencies,
            },
            // deeply nested resources have the same dependencies through parent inheritance
            DeeplyNestedResource_NestedCompositeResource1_NestedResource1_8B6D9004:
              {
                depends_on: expectedDependencies,
              },
            DeeplyNestedResource_NestedCompositeResource1_NestedResource2_42F5D27C:
              {
                depends_on: expectedDependencies,
              },
            DeeplyNestedResource_NestedPreSynthResource1_NestedResource1_8C8BB53A:
              {
                depends_on: expectedDependencies,
              },
          },
        },
      });
    });

    test("does not propagate nested dependency to siblings", () => {
      // GIVEN
      const resourceA = new TerraformResource(spec, "ResourceA", {
        terraformResourceType,
      });

      class CompositeWithNestedDependencyResource extends TerraformElement {
        constructor(scope: Construct, id: string) {
          super(scope, id);
          const nested1 = new TerraformResource(this, "NestedResource1", {
            terraformResourceType,
          });
          // nested Resource 1 should depend on resourceA
          nested1.node.addDependency(resourceA);
          // nested Resource 2 should not depend on resourceA
          new TerraformResource(this, "NestedResource2", {
            terraformResourceType,
          });
        }
      }
      // WHEN
      new CompositeWithNestedDependencyResource(spec, "ResourceB");

      // THEN
      const synthesized = Testing.synth(spec);
      // expect(synthesized).toMatchSnapshot();
      const template = JSON.parse(synthesized);
      expect(template).toMatchObject({
        resource: {
          test_resource: {
            ResourceB_NestedResource1_0872214E: {
              depends_on: ["test_resource.ResourceA"],
            },
            ResourceB_NestedResource2_477F69A1: expect.not.objectContaining({
              depends_on: expect.anything(),
            }),
          },
        },
      });
    });

    // TODO: Should throw circular dependency error during synth because TF sure will...
    test.skip("throws on circular dependencies", () => {
      // GIVEN
      const resourceA = new TerraformResource(spec, "ResourceA", {
        terraformResourceType,
      });
      const resourceB = new TerraformResource(spec, "ResourceB", {
        terraformResourceType,
      });

      // WHEN
      // Create circular dependency
      resourceA.node.addDependency(resourceB);
      expect(() => {
        resourceB.node.addDependency(resourceA);
      }).toThrow(/circular dependency/);

      // // THEN
      // spec.prepareStack();
      // const synthesized = Testing.synth(spec);
      // expect(synthesized).toMatchSnapshot();
      // const template = JSON.parse(synthesized);
      // expect(template).toMatchObject({
      //   resource: {
      //     [terraformResourceType]: {
      //       ResourceA: {
      //         depends_on: [`${terraformResourceType}.ResourceB`],
      //       },
      //       ResourceB: {
      //         depends_on: [`${terraformResourceType}.ResourceA`],
      //       },
      //     },
      //   },
      // });
    });
  });
});

class MySpec extends SpecBase {}
class CompositeResource extends TerraformElement {
  constructor(scope: Construct, id: string) {
    super(scope, id);
    new TerraformResource(this, "NestedResource1", {
      terraformResourceType,
    });
    new TerraformResource(this, "NestedResource2", {
      terraformResourceType,
    });
  }
}

class PreSynthResource extends TerraformElement {
  // additional resource added during prepareStack!
  public toTerraform(): any {
    const id = "NestedResource1";
    if (!this.node.tryFindChild(id)) {
      new TerraformResource(this, id, {
        terraformResourceType,
      });
    }
    return {};
  }
}
class DeeplyNestedResource extends TerraformElement {
  constructor(scope: Construct, id: string) {
    super(scope, id);
    new CompositeResource(this, "NestedCompositeResource1");
    new PreSynthResource(this, "NestedPreSynthResource1");
  }
}
// Helper function to add dependencies to multiple resources
function addDependencies(resources: any[], dependency: any) {
  resources.forEach((resource) => resource.node.addDependency(dependency));
}
