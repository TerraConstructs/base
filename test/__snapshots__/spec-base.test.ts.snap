// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpecBase TerraformDependencyAspect throws on circular dependencies 1`] = `
"{
  "resource": {
    "test_resource": {
      "ResourceA": {
        "depends_on": [
          "test_resource.ResourceB"
        ]
      },
      "ResourceB": {
        "depends_on": [
          "test_resource.ResourceA"
        ]
      }
    }
  },
  "terraform": {
    "backend": {
      "http": {
        "address": "http://localhost:3000"
      }
    }
  }
}"
`;
